File: Sources/VoxLocalCore/Audio/AudioRecorder.swift
import AVFoundation
import CoreAudio
import Foundation

public struct RecordingResult: Sendable {
    public let audioURL: URL
    public let duration: TimeInterval
    public let peakLevel: Float
}

/// Captures microphone audio with `AVAudioEngine`, converts it to
/// 16 kHz mono PCM16 on the fly and writes a temporary WAV file suitable
/// for whisper.cpp. The temporary file lives only for the duration of one
/// dictation session; callers must delete it via the returned URL (the
/// `DictationController` does) and `cancel()` deletes it immediately.
public final class AudioRecorder {
    public var levelHandler: ((Float) -> Void)?
    /// Called when the input device disappears mid-recording.
    public var deviceInterruptionHandler: (() -> Void)?

    private let engine = AVAudioEngine()
    // `lock` guards converter/writer/peak: process(buffer:) runs on the
    // audio tap thread while start/stop/cancel run on the main actor, and
    // removeTap does not guarantee in-flight callbacks have finished.
    private let lock = NSLock()
    private var converter: AVAudioConverter?
    private var writer: WAVWriter?
    private var peak: Float = 0
    private var observer: NSObjectProtocol?
    private let targetFormat = AVAudioFormat(
        commonFormat: .pcmFormatInt16, sampleRate: 16000, channels: 1, interleaved: true)!
    private(set) public var isRecording = false

    public init() {}

    public func start(deviceUID: String?) throws {
        guard !isRecording else { throw VoxLocalError.recordingFailed("already recording") }
        lock.lock()
        peak = 0
        lock.unlock()

        if let uid = deviceUID, let deviceID = AudioDeviceFinder.deviceID(forUID: uid) {
            try setInputDevice(deviceID)
        }

        let input = engine.inputNode
        let inputFormat = input.outputFormat(forBus: 0)
        guard inputFormat.sampleRate > 0, inputFormat.channelCount > 0 else {
            throw VoxLocalError.microphoneUnavailable
        }
        guard let converter = AVAudioConverter(from: inputFormat, to: targetFormat) else {
            throw VoxLocalError.recordingFailed("cannot build converter from \(inputFormat)")
        }

        let tempURL = FileManager.default.temporaryDirectory
            .appendingPathComponent("voxlocal-\(UUID().uuidString).wav")
        let writer = try WAVWriter(url: tempURL, sampleRate: 16000)
        lock.lock()
        self.converter = converter
        self.writer = writer
        lock.unlock()

        input.installTap(onBus: 0, bufferSize: 4096, format: inputFormat) { [weak self] buffer, _ in
            self?.process(buffer: buffer)
        }

        observer = NotificationCenter.default.addObserver(
            forName: .AVAudioEngineConfigurationChange, object: engine, queue: .main
        ) { [weak self] _ in
            guard let self, self.isRecording else { return }
            Log.shared.info("audio engine configuration changed mid-recording (device change/disconnect)")
            self.deviceInterruptionHandler?()
        }

        engine.prepare()
        do {
            try engine.start()
        } catch {
            cleanupEngine()
            lock.lock()
            self.writer = nil
            self.converter = nil
            lock.unlock()
            writer.cancelAndDelete()
            throw VoxLocalError.recordingFailed(error.localizedDescription)
        }
        isRecording = true
        Log.shared.info("recording started (input: \(Int(inputFormat.sampleRate)) Hz, \(inputFormat.channelCount) ch)")
    }

    private func process(buffer: AVAudioPCMBuffer) {
        // Entire body under the lock: stop()/cancel() nil out writer and
        // converter under the same lock, so no append can race finalize().
        lock.lock()
        defer { lock.unlock() }
        guard let converter, let writer else { return }

        // Level metering on the raw input buffer.
        if let ch = buffer.floatChannelData?[0] {
            let n = Int(buffer.frameLength)
            var sum: Float = 0
            var i = 0
            while i < n {
                let v = ch[i]
                sum += v * v
                i += 16 // sample every 16th frame; enough for a meter
            }
            let rms = sqrtf(sum / Float(max(1, n / 16)))
            let level = min(1.0, rms * 18)
            peak = max(peak, level)
            if let handler = levelHandler {
                DispatchQueue.main.async { handler(level) }
            }
        }

        let ratio = targetFormat.sampleRate / buffer.format.sampleRate
        let capacity = AVAudioFrameCount(Double(buffer.frameLength) * ratio) + 64
        guard let out = AVAudioPCMBuffer(pcmFormat: targetFormat, frameCapacity: capacity) else { return }
        var supplied = false
        var convError: NSError?
        let status = converter.convert(to: out, error: &convError) { _, outStatus in
            if supplied {
                outStatus.pointee = .noDataNow
                return nil
            }
            supplied = true
            outStatus.pointee = .haveData
            return buffer
        }
        guard status != .error, convError == nil else {
            Log.shared.error("audio conversion failed: \(convError?.localizedDescription ?? "unknown")")
            return
        }
        if out.frameLength > 0, let samples = out.int16ChannelData?[0] {
            try? writer.append(samples, count: Int(out.frameLength))
        }
    }

    /// Stops the engine, finalizes the WAV and returns it. Throws
    /// `.emptyRecording` (and deletes the file) when almost nothing was
    /// captured — e.g. a too-short press or a dead microphone.
    public func stop() throws -> RecordingResult {
        guard isRecording else { throw VoxLocalError.recordingFailed("not recording") }
        cleanupEngine()
        isRecording = false

        // Detach shared state under the lock; any in-flight tap callback
        // either finished before this or sees nil and bails out.
        lock.lock()
        let writer = self.writer
        self.writer = nil
        self.converter = nil
        let capturedPeak = peak
        lock.unlock()
        guard let writer else { throw VoxLocalError.recordingFailed("not recording") }

        let duration = writer.duration
        // Below ~0.35 s or near-silence there is nothing Whisper can use.
        if duration < 0.35 || capturedPeak < 0.005 {
            writer.cancelAndDelete()
            Log.shared.info("recording rejected as empty (duration \(String(format: "%.2f", duration))s, peak \(String(format: "%.3f", capturedPeak)))")
            throw VoxLocalError.emptyRecording
        }
        try writer.finalize()
        Log.shared.info("recording stopped (\(String(format: "%.2f", duration))s)")
        return RecordingResult(audioURL: writer.url, duration: duration, peakLevel: capturedPeak)
    }

    /// Stops immediately and deletes the temporary audio.
    public func cancel() {
        lock.lock()
        let writer = self.writer
        self.writer = nil
        self.converter = nil
        lock.unlock()
        guard isRecording || writer != nil else { return }
        cleanupEngine()
        isRecording = false
        writer?.cancelAndDelete()
        Log.shared.info("recording cancelled; temp audio deleted")
    }

    private func cleanupEngine() {
        engine.inputNode.removeTap(onBus: 0)
        engine.stop()
        if let observer {
            NotificationCenter.default.removeObserver(observer)
            self.observer = nil
        }
    }

    private func setInputDevice(_ deviceID: AudioDeviceID) throws {
        guard let unit = engine.inputNode.audioUnit else {
            throw VoxLocalError.microphoneUnavailable
        }
        var device = deviceID
        let status = AudioUnitSetProperty(
            unit,
            kAudioOutputUnitProperty_CurrentDevice,
            kAudioUnitScope_Global,
            0,
            &device,
            UInt32(MemoryLayout<AudioDeviceID>.size))
        if status != noErr {
            Log.shared.error("failed to select input device (OSStatus \(status)); falling back to default")
        }
    }
}

/// CoreAudio lookups for input-device selection in Settings.
public enum AudioDeviceFinder {
    public struct Device: Identifiable, Hashable, Sendable {
        public let uid: String
        public let name: String
        public var id: String { uid }
    }

    public static func inputDevices() -> [Device] {
        var address = AudioObjectPropertyAddress(
            mSelector: kAudioHardwarePropertyDevices,
            mScope: kAudioObjectPropertyScopeGlobal,
            mElement: kAudioObjectPropertyElementMain)
        var size: UInt32 = 0
        guard AudioObjectGetPropertyDataSize(AudioObjectID(kAudioObjectSystemObject), &address, 0, nil, &size) == noErr else {
            return []
        }
        let count = Int(size) / MemoryLayout<AudioDeviceID>.size
        var ids = [AudioDeviceID](repeating: 0, count: count)
        guard AudioObjectGetPropertyData(AudioObjectID(kAudioObjectSystemObject), &address, 0, nil, &size, &ids) == noErr else {
            return []
        }
        return ids.compactMap { id in
            guard inputChannelCount(id) > 0, let uid = stringProperty(id, kAudioDevicePropertyDeviceUID),
                  let name = stringProperty(id, kAudioObjectPropertyName) else { return nil }
            return Device(uid: uid, name: name)
        }
    }

    public static func deviceID(forUID uid: String) -> AudioDeviceID? {
        var address = AudioObjectPropertyAddress(
            mSelector: kAudioHardwarePropertyDevices,
            mScope: kAudioObjectPropertyScopeGlobal,
            mElement: kAudioObjectPropertyElementMain)
        var size: UInt32 = 0
        guard AudioObjectGetPropertyDataSize(AudioObjectID(kAudioObjectSystemObject), &address, 0, nil, &size) == noErr else {
            return nil
        }
        let count = Int(size) / MemoryLayout<AudioDeviceID>.size
        var ids = [AudioDeviceID](repeating: 0, count: count)
        guard AudioObjectGetPropertyData(AudioObjectID(kAudioObjectSystemObject), &address, 0, nil, &size, &ids) == noErr else {
            return nil
        }
        return ids.first { stringProperty($0, kAudioDevicePropertyDeviceUID) == uid }
    }

    private static func inputChannelCount(_ id: AudioDeviceID) -> Int {
        var address = AudioObjectPropertyAddress(
            mSelector: kAudioDevicePropertyStreamConfiguration,
            mScope: kAudioDevicePropertyScopeInput,
            mElement: kAudioObjectPropertyElementMain)
        var size: UInt32 = 0
        guard AudioObjectGetPropertyDataSize(id, &address, 0, nil, &size) == noErr, size > 0 else { return 0 }
        let bufferList = UnsafeMutablePointer<AudioBufferList>.allocate(capacity: Int(size))
        defer { bufferList.deallocate() }
        guard AudioObjectGetPropertyData(id, &address, 0, nil, &size, bufferList) == noErr else { return 0 }
        let buffers = UnsafeMutableAudioBufferListPointer(bufferList)
        return buffers.reduce(0) { $0 + Int($1.mNumberChannels) }
    }

    private static func stringProperty(_ id: AudioDeviceID, _ selector: AudioObjectPropertySelector) -> String? {
        var address = AudioObjectPropertyAddress(
            mSelector: selector,
            mScope: kAudioObjectPropertyScopeGlobal,
            mElement: kAudioObjectPropertyElementMain)
        var cfString: Unmanaged<CFString>?
        var size = UInt32(MemoryLayout<Unmanaged<CFString>?>.size)
        let status = withUnsafeMutablePointer(to: &cfString) { ptr in
            AudioObjectGetPropertyData(id, &address, 0, nil, &size, ptr)
        }
        guard status == noErr, let value = cfString?.takeRetainedValue() else { return nil }
        return value as String
    }
}

File: windows/src/VoxLocal.App/Core/Audio/AudioRecorder.cs
using NAudio.CoreAudioApi;
using NAudio.Wave;
using NAudio.Wave.SampleProviders;
using System.Runtime.InteropServices;
using VoxLocal.Core.Models;
using VoxLocal.Core.Utilities;

namespace VoxLocal.Core.Audio;

public sealed record RecordingResult(string AudioPath, TimeSpan Duration, float PeakLevel);

/// <summary>
/// Captures microphone audio with WASAPI (NAudio), converts it to
/// 16 kHz mono PCM16 on the fly and writes a temporary WAV file suitable
/// for whisper.cpp. The temporary file lives only for the duration of one
/// dictation session; callers must delete it via the returned path (the
/// DictationController does) and Cancel() deletes it immediately.
/// </summary>
public sealed class AudioRecorder
{
    public Action<float>? LevelHandler { get; set; }
    /// <summary>Called when the input device disappears mid-recording.</summary>
    public Action? DeviceInterruptionHandler { get; set; }

    // _lock guards writer/pipeline/peak: OnDataAvailable runs on the WASAPI
    // capture thread while Start/Stop/Cancel run on the UI thread.
    private readonly object _lock = new();
    private WasapiCapture? _capture;
    private BufferedWaveProvider? _sourceBuffer;
    private IWaveProvider? _pcm16Pipeline;
    private WavWriter? _writer;
    private float _peak;
    private bool _stopping;
    private readonly byte[] _drainBuffer = new byte[16000 * 2]; // 1 s of PCM16 mono

    public bool IsRecording { get; private set; }

    public void Start(string? deviceId)
    {
        if (IsRecording)
        {
            throw new VoxLocalException(new VoxLocalError.RecordingFailed("already recording"));
        }
        lock (_lock) { _peak = 0; }

        WasapiCapture capture;
        try
        {
            var device = deviceId is null ? null : AudioDeviceFinder.Find(deviceId);
            capture = device is null ? new WasapiCapture() : new WasapiCapture(device);
        }
        catch (Exception)
        {
            throw new VoxLocalException(new VoxLocalError.MicrophoneUnavailable());
        }

        var format = capture.WaveFormat;
        if (format.SampleRate <= 0 || format.Channels <= 0)
        {
            capture.Dispose();
            throw new VoxLocalException(new VoxLocalError.MicrophoneUnavailable());
        }

        var tempPath = Path.Combine(Path.GetTempPath(), $"voxlocal-{Guid.NewGuid()}.wav");
        var writer = new WavWriter(tempPath, 16000);

        // Conversion pipeline: source buffer → mono → resample to 16 kHz → PCM16.
        var sourceBuffer = new BufferedWaveProvider(format)
        {
            DiscardOnBufferOverflow = true,
            ReadFully = false,
        };
        ISampleProvider samples = sourceBuffer.ToSampleProvider();
        if (format.Channels > 1)
        {
            // Take channel 0; microphones are effectively mono sources.
            samples = new MultiplexingSampleProvider(new[] { samples }, 1);
        }
        var resampled = new WdlResamplingSampleProvider(samples, 16000);
        var pcm16 = resampled.ToWaveProvider16();

        capture.DataAvailable += OnDataAvailable;
        capture.RecordingStopped += OnRecordingStopped;

        lock (_lock)
        {
            _capture = capture;
            _sourceBuffer = sourceBuffer;
            _pcm16Pipeline = pcm16;
            _writer = writer;
            _stopping = false;
        }

        try
        {
            capture.StartRecording();
        }
        catch (Exception ex)
        {
            CleanupCapture();
            lock (_lock)
            {
                _writer = null;
                _sourceBuffer = null;
                _pcm16Pipeline = null;
            }
            writer.CancelAndDelete();
            throw new VoxLocalException(new VoxLocalError.RecordingFailed(ex.Message));
        }
        IsRecording = true;
        Log.Shared.Info($"recording started (input: {format.SampleRate} Hz, {format.Channels} ch)");
    }

    private void OnDataAvailable(object? sender, WaveInEventArgs e)
    {
        // Entire body under the lock: Stop()/Cancel() null out the writer and
        // pipeline under the same lock, so no append can race Complete().
        lock (_lock)
        {
            if (_writer is null || _sourceBuffer is null || _pcm16Pipeline is null || _capture is null)
            {
                return;
            }

            // Level metering on the raw input buffer.
            var level = ComputeLevel(e.Buffer, e.BytesRecorded, _capture.WaveFormat);
            _peak = Math.Max(_peak, level);
            if (LevelHandler is { } handler)
            {
                System.Windows.Application.Current?.Dispatcher.BeginInvoke(() => handler(level));
            }

            _sourceBuffer.AddSamples(e.Buffer, 0, e.BytesRecorded);

            // Drain everything currently convertible into the WAV file.
            int read;
            while ((read = _pcm16Pipeline.Read(_drainBuffer, 0, _drainBuffer.Length)) > 0)
            {
                _writer.Append(_drainBuffer.AsSpan(0, read));
            }
        }
    }

    /// <summary>
    /// Perceptual meter level: sampled RMS mapped through a decibel curve
    /// (-50 dB … -8 dB → 0 … 1). Normal conversational speech lands around
    /// the middle of the meter instead of barely lighting two bars — the old
    /// linear "RMS × 18" heuristic required shouting to move the meter.
    /// </summary>
    private static float ComputeLevel(byte[] buffer, int bytes, WaveFormat format)
    {
        var step = 16 * format.Channels; // sample every 16th frame; enough for a meter
        float rms;
        if (format.Encoding == WaveFormatEncoding.IeeeFloat && bytes >= 4)
        {
            var floats = MemoryMarshal.Cast<byte, float>(buffer.AsSpan(0, bytes));
            float sum = 0;
            var counted = 0;
            for (var i = 0; i < floats.Length; i += step)
            {
                sum += floats[i] * floats[i];
                counted++;
            }
            rms = MathF.Sqrt(sum / Math.Max(1, counted));
        }
        else if (format.Encoding == WaveFormatEncoding.Pcm && format.BitsPerSample == 16 && bytes >= 2)
        {
            var shorts = MemoryMarshal.Cast<byte, short>(buffer.AsSpan(0, bytes));
            float sum = 0;
            var counted = 0;
            for (var i = 0; i < shorts.Length; i += step)
            {
                var v = shorts[i] / 32768f;
                sum += v * v;
                counted++;
            }
            rms = MathF.Sqrt(sum / Math.Max(1, counted));
        }
        else
        {
            return 0;
        }

        if (rms <= 0.0001f)
        {
            return 0;
        }
        var db = 20f * MathF.Log10(rms);
        return Math.Clamp((db + 50f) / 42f, 0f, 1f);
    }

    private void OnRecordingStopped(object? sender, StoppedEventArgs e)
    {
        if (e.Exception is not null && IsRecording && !_stopping)
        {
            Log.Shared.Info("audio capture stopped unexpectedly mid-recording (device change/disconnect)");
            DeviceInterruptionHandler?.Invoke();
        }
    }

    /// <summary>
    /// Stops the capture, finalizes the WAV and returns it. Throws
    /// EmptyRecording (and deletes the file) when almost nothing was
    /// captured — e.g. a too-short press or a dead microphone.
    /// </summary>
    public RecordingResult Stop()
    {
        if (!IsRecording)
        {
            throw new VoxLocalException(new VoxLocalError.RecordingFailed("not recording"));
        }
        lock (_lock) { _stopping = true; }
        CleanupCapture();
        IsRecording = false;

        WavWriter? writer;
        float capturedPeak;
        lock (_lock)
        {
            writer = _writer;
            _writer = null;
            _sourceBuffer = null;
            _pcm16Pipeline = null;
            capturedPeak = _peak;
        }
        if (writer is null)
        {
            throw new VoxLocalException(new VoxLocalError.RecordingFailed("not recording"));
        }

        var duration = writer.Duration;
        // Below ~0.35 s or near-silence there is nothing Whisper can use.
        if (duration.TotalSeconds < 0.35 || capturedPeak < 0.005f)
        {
            writer.CancelAndDelete();
            Log.Shared.Info($"recording rejected as empty (duration {duration.TotalSeconds:F2}s, peak {capturedPeak:F3})");
            throw new VoxLocalException(new VoxLocalError.EmptyRecording());
        }
        writer.Complete();
        Log.Shared.Info($"recording stopped ({duration.TotalSeconds:F2}s)");
        return new RecordingResult(writer.Path, duration, capturedPeak);
    }

    /// <summary>Stops immediately and deletes the temporary audio.</summary>
    public void Cancel()
    {
        WavWriter? writer;
        lock (_lock)
        {
            _stopping = true;
            writer = _writer;
            _writer = null;
            _sourceBuffer = null;
            _pcm16Pipeline = null;
        }
        if (!IsRecording && writer is null) return;
        CleanupCapture();
        IsRecording = false;
        writer?.CancelAndDelete();
        Log.Shared.Info("recording cancelled; temp audio deleted");
    }

    private void CleanupCapture()
    {
        WasapiCapture? capture;
        lock (_lock)
        {
            capture = _capture;
            _capture = null;
        }
        if (capture is null) return;
        capture.DataAvailable -= OnDataAvailable;
        capture.RecordingStopped -= OnRecordingStopped;
        try { capture.StopRecording(); } catch { /* not started */ }
        capture.Dispose();
    }
}

/// <summary>WASAPI device lookups for input-device selection in Settings.</summary>
public static class AudioDeviceFinder
{
    public sealed record Device(string Id, string Name);

    public static IReadOnlyList<Device> InputDevices()
    {
        using var enumerator = new MMDeviceEnumerator();
        var devices = new List<Device>();
        foreach (var device in enumerator.EnumerateAudioEndPoints(DataFlow.Capture, DeviceState.Active))
        {
            using (device)
            {
                devices.Add(new Device(device.ID, device.FriendlyName));
            }
        }
        return devices;
    }

    internal static MMDevice? Find(string id)
    {
        using var enumerator = new MMDeviceEnumerator();
        try
        {
            var device = enumerator.GetDevice(id);
            return device is { State: DeviceState.Active } ? device : null;
        }
        catch
        {
            return null;
        }
    }
}
